tools/codegen/context.py

```diff
@@ -8,19 +9,33 @@
 
 # Helper functions for defining generators on things in the model
 
-F = TypeVar('F', NativeFunction, NativeFunctionsGroup, Union[NativeFunction, NativeFunctionsGroup])
+F = TypeVar(
+    'F',
+    NativeFunction,
+    NativeFunctionsGroup,
+    ExternalBackendFunction,
+    ExternalBackendFunctionsGroup,
+    Union[NativeFunction, NativeFunctionsGroup],
+    Union[ExternalBackendFunctionsGroup, ExternalBackendFunction],
+    Union[NativeFunction, NativeFunctionsGroup, ExternalBackendFunction, ExternalBackendFunctionsGroup]
+)
 
 @contextlib.contextmanager
-def native_function_manager(g: Union[NativeFunctionsGroup, NativeFunction]) -> Iterator[None]:
-    if isinstance(g, NativeFunctionsGroup):
+def native_function_manager(g: Union[
+        NativeFunctionsGroup, NativeFunction, ExternalBackendFunction, ExternalBackendFunctionsGroup]) -> Iterator[None]:
+    if isinstance(g, ExternalBackendFunctionsGroup):
+        f = g.primary.native_function
+    elif isinstance(g, ExternalBackendFunction):
+        f = g.native_function
+    elif isinstance(g, NativeFunctionsGroup):
         # By default, we associate all errors with structured native functions
         # with the out variant.  In some cases, it might be better to have
         # a more specific place to hang things; if so, use
         # native_function_manager again on the inside
         f = g.out
     else:
         f = g
-    with context(f'in {f.loc}:\n  {f.func}'):
+    with context(f'in native_functions.yaml line {f.loc}:\n  {f.func}'):
```

Comment:
I'm not sure there's any simple way to make the file name less hardcoded here, but it would be nice. E.g. if file loading in gen.py and this here both referred to a global or something. Not a huge deal but would forestall drift later etc.